pkg/virt-launcher/virtwrap/network/common.go

```diff
@@ -406,24 +400,35 @@ func buildTapDeviceMaker(tapName string, isMultiqueue bool, virtLauncherPID int)
 		createTapDeviceArgs = append(createTapDeviceArgs, "--multiqueue")
 	}
 	cmd := exec.Command("virt-chroot", createTapDeviceArgs...)
+
 	return &tapDeviceMaker{
-		virtLauncherSELinuxLabel: virtLauncherSELinuxLabel,
-		virtHandlerSELinuxLabel:  virtHandlerSELinuxLabel,
-		tapMakingCmd:             cmd,
-		tapName:                  tapName,
-		isMultiqueue:             isMultiqueue,
+		tapMakingCmd:    cmd,
+		tapName:         tapName,
+		isMultiqueue:    isMultiqueue,
+		virtLauncherPID: virtLauncherPID,
 	}, nil
 }
 
 func (tmc *tapDeviceMaker) makeTapDevice() error {
-	runtime.LockOSThread()
-	defer func() {
-		_ = selinux.SetExecLabel(tmc.virtHandlerSELinuxLabel)
-		runtime.UnlockOSThread()
-	}()
+	virtHandlerSELinuxLabel := "system_u:system_r:spc_t:s0"
+	virtLauncherSELinuxLabel := ""
+
+	_, selinuxEnabled, err := kvselinux.NewSELinux()
+	if err == nil && selinuxEnabled {
+		virtLauncherSELinuxLabel, err = getProcessCurrentSELinuxLabel(tmc.virtLauncherPID)
+		if err != nil {
+			return fmt.Errorf("error reading virt-launcher %d selinux label. Reason: %v", tmc.virtLauncherPID, err)
+		}
+
+		runtime.LockOSThread()
+		defer func() {
+			_ = selinux.SetExecLabel(virtHandlerSELinuxLabel)
+			runtime.UnlockOSThread()
+		}()
 
-	if err := selinux.SetExecLabel(tmc.virtLauncherSELinuxLabel); err != nil {
-		return fmt.Errorf("failed to switch selinux context to %s. Reason: %v", tmc.virtLauncherSELinuxLabel, err)
+		if err := selinux.SetExecLabel(virtLauncherSELinuxLabel); err != nil {
+			return fmt.Errorf("failed to switch selinux context to %s. Reason: %v", virtLauncherSELinuxLabel, err)
+		}
 	}
```

Comment:
Please use dedicated functions here to check if selinux is enabled and to relabel if needed. This "optional" code should not take up more than half of the function called `makeTapDevice`. Hide SELinux for the casual reader.

Reply:
`makeTapDevice` was 'born' exactly for that reason: hide selinux from the user; creating the tap device, as you've identified, is just executing a command. Now, we've learned that sometimes, you must stay away from it.

I feel adding yet another layer is both unnecessary (for the reasons explained above), and would further complicate this, since I would have to stop relying on the `defer` to maker sure that when needed I reset the selinux context back to the original one.

Reply:
The original version of this function is quite clear. Despite it contains SELinux handling, it is just a call to two functions without any variables/constants/logic. All I need to read there is "set SELinux now and reset it at the end".

The following function `preventFDLeakOntoChild` stands as an example that there was an intention to hide implementation details from this body.

Your change introduces not only the `if`, which would be minor, it also:
- Initializes SELinuxClient instance to **only** check whether SELinux is available. On this level, it looks like an abuse of the unused client
- Prepare label variables that are not used for anything but configuration of SELinux inside this `if` block (can't you move it there?)

I suggest you at least create a helper checking for selinux being enabled and move the variable, or better (since the SELinux portion of the body grew), abstract even the rest of the code as following:

```
func makeTapDevice()
  if selinuxEnabled()
    err = setLabel()
    defer resetLabel()
    if err
      exit

  non selinux content
```

Reply:
> The original version of this function is quite clear. Despite it contains SELinux handling, it is just a call to two functions without any variables/constants/logic. All I need to read there is "set SELinux now and reset it at the end".
> 
> The following function `preventFDLeakOntoChild` stands as an example that there was an intention to hide implementation details from this body.
> 
> Your change introduces not only the `if`, which would be minor, it also:
> 
> * Initializes SELinuxClient instance to **only** check whether SELinux is available. On this level, it looks like an abuse of the unused client

I cannot help but point out that the **only** you're upset about here is implemented by the **if** that you're saying is minor above. 

> * Prepare label variables that are not used for anything but configuration of SELinux inside this `if` block (can't you move it there?)

You're right here. Will move those.

> 
> I suggest you at least create a helper checking for selinux being enabled and move the variable, or better (since the SELinux portion of the body grew), abstract even the rest of the code as following:
> 
> ```
> func makeTapDevice()
>   if selinuxEnabled()
>     err = setLabel()
>     defer resetLabel()
>     if err
>       exit
> 
>   non selinux content
> ```